---
title: "Analysis of Sequence Variants in Clones Generated from ME180 Cells by Genome Editing"
author: "Anna Hahut and Jelena Popovic"
date: "`r Sys.Date()`"
output: 
  html_document: 
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Working Directory
```{r}
setwd("C:/Users/xwl2686/OneDrive - Northwestern University/Desktop/GitHub")
```

# Introduction to R Environment and Setup

Data were analyzed using R version 4.3.2. We utilized the `readxl`, `dplyr`, `stringr`, and `tidyverse` libraries to read, transform, and clean the data for subsequent analysis. For data visualization, we used the `gt`, `ggVennDiagram`, `circlize`, `ggplot2`, `gridExtra` and `grid` libraries. 

Loading required packages
```{r echo=TRUE, message=FALSE, warning=FALSE}

library(tidyverse)
library(readxl)
library(dplyr)
library(stringr)
library(gt)
library(circlize)
library(ggVennDiagram)
library(ggplot2)
library(gridExtra)
library(grid)
```

# Data Preparation

The following R code was used to import and preprocess DNA sequence data of CRISPR/Cas9 mutant clones compared to ME180 WT cells. 
The datasets were imported from Excel files using the `read_excel()` function. 
The data was organized using column mutations. New columns (CHROM_POS, CONTQ, DP, TLOD, and ANN) were created by concatenating and extracting different specific information from the INFO column.The content of the INFO cells includes details about the gene mutation that is detected: the effect of the variant (stop loss, stop gain, etc.), effect impact on gene function (High, Moderate, Low or Modifier), functional class of the variant (nonsense, missense, frameshift etc.), codon change, amino acid change, amino acid chain length, gene name, gene biotype (protein coding, pseudogene…), coding information, transcript information, exon information and any errors or warnings detected. The effect impact is what SnpEff uses to determine how deleterious the variant is with regard to the function of a gene. For example, a HIGH impact output means that SnpEff predicts that the variant causes deleterious gene effects [Input & output files](https://pcingola.github.io/SnpEff/snpeff/inputoutput/).
The number of transcripts in the ANN column was counted by counting the number of commas +1. The "ANN" column was separated into multiple rows based on commas and then split still further into multiple columns. The relevant columns were selected for further analysis, including renamed columns. Empty entries in the "Biotype" column were replaced with "noncoding". Empty entries in the "Feature_Type" column were replaced with "NA". 

#### VII11

```{r}
df3VII11<-read_excel("DNA DATA of CRISPR clones compared to ME180.xlsx",
                     sheet="VII11")|>
  as.data.frame()|>
  mutate("CLONE"="VII11")|>
  relocate(CLONE,.before = CHROM)|>
  mutate("CHROM_POS"=paste(CHROM,POS,sep="="))|>
  mutate(CONTQ = gsub(".*CONTQ=([^;]+).*", "\\1", INFO),# Extracting data from the INFO column and placing it into appropriate columns
         DP = gsub(".*DP=([^;]+).*", "\\1", INFO),
         TLOD = gsub(".*TLOD=([^;]+).*", "\\1", INFO),
         ANN = gsub(".*ANN=([^;]+).*", "\\1", INFO)
  )|>
  mutate(transcripts_N = str_count(ANN, ",")+1)|>  # Counting the number of parts separated by commas in the ANN column
  separate_rows(ANN, sep = ",")|>  #To separate the contents of the ANN column into different rows based on commas
  separate(col=ANN,into=LETTERS[seq(from=1,to=16)],sep="\\|",remove=F)|> # Extracting data from the ANN
  select("CLONE","CHROM_POS","CHROM","POS","REF","ALT","INFO","CONTQ","DP","TLOD","transcripts_N","ANN", "Allele"=A,"Effect"=B,"Effect_Impact"=C,"Gene"=D,"Gene_name"=E,
         "Feature_Type"="F","Feature_ID"=G,"Biotype"=H,"Rank"=I,"HGVS_cDNA"=J,"HGVS_Protein"=K,"Distance"=L,"Distance1"=M,"Distance2"=N,"Distance3"=O,"Errors_Warnings_INFO"=P,"FORMAT")|>
  mutate(Biotype = if_else(Biotype == "", "noncoding", Biotype))|>
  mutate(Feature_Type = if_else(Feature_Type == "", "NA", Feature_Type))
```

#### A13

```{r}
df3A13<-read_excel("DNA DATA of CRISPR clones compared to ME180.xlsx",
                   sheet="A13")|>
  as.data.frame()|>
  mutate("CLONE"="A13")|>
  relocate(CLONE,.before = CHROM)|>
  mutate("CHROM_POS"=paste(CHROM,POS,sep="="))|>
  mutate(CONTQ = gsub(".*CONTQ=([^;]+).*", "\\1", INFO),# Extracting data from the INFO column and placing it into appropriate columns
         DP = gsub(".*DP=([^;]+).*", "\\1", INFO),
         TLOD = gsub(".*TLOD=([^;]+).*", "\\1", INFO),
         ANN = gsub(".*ANN=([^;]+).*", "\\1", INFO)
  )|>
  mutate(transcripts_N = str_count(ANN, ",")+1)|>  # Counting the number of parts separated by commas in the ANN column
  separate_rows(ANN, sep = ",")|>  #To separate the contents of the ANN column into different rows based on commas
  separate(col=ANN,into=LETTERS[seq(from=1,to=16)],sep="\\|",remove=F)|> # Extracting data from the ANN
  select("CLONE","CHROM_POS","CHROM","POS","REF","ALT","INFO","CONTQ","DP","TLOD","transcripts_N","ANN", "Allele"=A,"Effect"=B,"Effect_Impact"=C,"Gene"=D,"Gene_name"=E,
         "Feature_Type"="F","Feature_ID"=G,"Biotype"=H,"Rank"=I,"HGVS_cDNA"=J,"HGVS_Protein"=K,"Distance"=L,"Distance1"=M,"Distance2"=N,"Distance3"=O,"Errors_Warnings_INFO"=P,"FORMAT")|>
  mutate(Biotype = if_else(Biotype == "", "noncoding", Biotype))|>
  mutate(Feature_Type = if_else(Feature_Type == "", "NA", Feature_Type))
```

#### A40

```{r}
df3A40<-read_excel("DNA DATA of CRISPR clones compared to ME180.xlsx",sheet="A40")|>
  as.data.frame()|>
  mutate("CLONE"="A40")|>
  relocate(CLONE,.before = CHROM)|>
  mutate("CHROM_POS"=paste(CHROM,POS,sep="="))|>
  mutate(CONTQ = gsub(".*CONTQ=([^;]+).*", "\\1", INFO),# Extracting data from the INFO column and placing it into appropriate columns
         DP = gsub(".*DP=([^;]+).*", "\\1", INFO),
         TLOD = gsub(".*TLOD=([^;]+).*", "\\1", INFO),
         ANN = gsub(".*ANN=([^;]+).*", "\\1", INFO)
  )|>
  mutate(transcripts_N = str_count(ANN, ",")+1)|>  # Counting the number of parts separated by commas in the ANN column
  separate_rows(ANN, sep = ",")|>  #To separate the contents of the ANN column into different rows based on commas
  separate(col=ANN,into=LETTERS[seq(from=1,to=16)],sep="\\|",remove=F)|> # Extracting data from the ANN
  select("CLONE","CHROM_POS","CHROM","POS","REF","ALT","INFO","CONTQ","DP","TLOD","transcripts_N","ANN", "Allele"=A,"Effect"=B,"Effect_Impact"=C,"Gene"=D,"Gene_name"=E,
         "Feature_Type"="F","Feature_ID"=G,"Biotype"=H,"Rank"=I,"HGVS_cDNA"=J,"HGVS_Protein"=K,"Distance"=L,"Distance1"=M,"Distance2"=N,"Distance3"=O,"Errors_Warnings_INFO"=P,"FORMAT")|>
  mutate(Biotype = if_else(Biotype == "", "noncoding", Biotype))|>
  mutate(Feature_Type = if_else(Feature_Type == "", "NA", Feature_Type))
```

#### II21

```{r}
df3II21<-read_excel("DNA DATA of CRISPR clones compared to ME180.xlsx",
                    sheet="II21")|>
  as.data.frame()|>
  mutate("CLONE"="II21")|>
  relocate(CLONE,.before = CHROM)|>
  mutate("CHROM_POS"=paste(CHROM,POS,sep="="))|>
  mutate(CONTQ = gsub(".*CONTQ=([^;]+).*", "\\1", INFO),# Extracting data from the INFO column and placing it into appropriate columns
         DP = gsub(".*DP=([^;]+).*", "\\1", INFO),
         TLOD = gsub(".*TLOD=([^;]+).*", "\\1", INFO),
         ANN = gsub(".*ANN=([^;]+).*", "\\1", INFO)
  )|>
  mutate(transcripts_N = str_count(ANN, ",")+1)|>  # Counting the number of parts separated by commas in the ANN column
  separate_rows(ANN, sep = ",")|>  #To separate the contents of the ANN column into different rows based on commas
  separate(col=ANN,into=LETTERS[seq(from=1,to=16)],sep="\\|",remove=F)|> # Extracting data from the ANN
  select("CLONE","CHROM_POS","CHROM","POS","REF","ALT","INFO","CONTQ","DP","TLOD","transcripts_N","ANN", "Allele"=A,"Effect"=B,"Effect_Impact"=C,"Gene"=D,"Gene_name"=E,
         "Feature_Type"="F","Feature_ID"=G,"Biotype"=H,"Rank"=I,"HGVS_cDNA"=J,"HGVS_Protein"=K,"Distance"=L,"Distance1"=M,"Distance2"=N,"Distance3"=O,"Errors_Warnings_INFO"=P,"FORMAT")|>
  mutate(Biotype = if_else(Biotype == "", "noncoding", Biotype))|>
  mutate(Feature_Type = if_else(Feature_Type == "", "NA", Feature_Type))

```

#### IX44

```{r}
df3IX44<-read_excel("DNA DATA of CRISPR clones compared to ME180.xlsx",
                    sheet="IX44")|>
  as.data.frame()|>
  mutate("CLONE"="IX44")|>
  relocate(CLONE,.before = CHROM)|>
  mutate("CHROM_POS"=paste(CHROM,POS,sep="="))|>
  mutate(CONTQ = gsub(".*CONTQ=([^;]+).*", "\\1", INFO),# Extracting data from the INFO column and placing it into appropriate columns
         DP = gsub(".*DP=([^;]+).*", "\\1", INFO),
         TLOD = gsub(".*TLOD=([^;]+).*", "\\1", INFO),
         ANN = gsub(".*ANN=([^;]+).*", "\\1", INFO)
  )|>
  mutate(transcripts_N = str_count(ANN, ",")+1)|>  # Counting the number of parts separated by commas in the ANN column
  separate_rows(ANN, sep = ",")|>  #To separate the contents of the ANN column into different rows based on commas
  separate(col=ANN,into=LETTERS[seq(from=1,to=16)],sep="\\|",remove=F)|> # Extracting data from the ANN
  select("CLONE","CHROM_POS","CHROM","POS","REF","ALT","INFO","CONTQ","DP","TLOD","transcripts_N","ANN", "Allele"=A,"Effect"=B,"Effect_Impact"=C,"Gene"=D,"Gene_name"=E,
         "Feature_Type"="F","Feature_ID"=G,"Biotype"=H,"Rank"=I,"HGVS_cDNA"=J,"HGVS_Protein"=K,"Distance"=L,"Distance1"=M,"Distance2"=N,"Distance3"=O,"Errors_Warnings_INFO"=P,"FORMAT")|>
  mutate(Biotype = if_else(Biotype == "", "noncoding", Biotype))|>
  mutate(Feature_Type = if_else(Feature_Type == "", "NA", Feature_Type))
```

The data were aggregated into a single table and summarized by the column labeled "ALL" to obtain unique detailed annotations. Subsequently, the resulting data frame was saved as a CSV file using the  `write.csv()` function.
```{r}
all_codingSnpEffCLONES<-as_tibble(rbind(df3VII11,df3A13,df3A40,df3II21,df3IX44))%>%
  mutate("ALL"=paste(Biotype,CLONE,CHROM,POS,REF,ALT,CONTQ,DP,TLOD,transcripts_N,Allele,Effect,Effect_Impact,Gene,Gene_name,
                     Feature_Type,sep="="))%>%
  select("CHROM_POS","ALL","Feature_ID","Rank","HGVS_cDNA","HGVS_Protein","Distance","Distance1","Distance2","Distance3","Errors_Warnings_INFO","FORMAT")

```

```{r}
codingSnpEffCLONES1<-all_codingSnpEffCLONES%>%
  group_by(ALL)|>
  summarise_all(~toString(unique(.)))
```

```{r}
all_codingSnpEffCLONES2<-separate(codingSnpEffCLONES1,col=ALL,into=c("Biotype","CLONE","CHROM","POS","REF","ALT","CONTQ","DP","TLOD","transcripts_N","Allele","Effect","Effect_Impact","GENE","Gene_name","Feature_Type"),sep="=",remove=T)
```

Finally, two data frames were created: "Clones" and "Clones1". The "Clones" data frame contained information for each unique genetic variant, with each row representing a specific variant. The "Clones1" data frame, on the other hand, included information for each unique variant position. A single genomic position can be associated with multiple genes due to various factors such as overlapping genes, alternative splicing events, and the presence of regulatory elements. In genomic studies, a position refers to a specific location within the genome. Any specific position  may fall within the coding region or regulatory region of several genes, particularly in regions of high gene density or where genes are closely spaced [(Wright, Molloy, & Jaschke, 2022)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8490965/).

```{r}
Clones<-all_codingSnpEffCLONES2%>%
  mutate(CHROM_POS_Clone = paste(CHROM_POS, CLONE, sep = "="))
```

```{r}
Clones1<-Clones %>%
  group_by(CHROM_POS_Clone) %>%
  summarise_all(~toString(unique(.))) 
```

# Visualization 

### Total number of genomic sequence variants in clones that were detected as different from genome of wild type ME180 cells

The total count of genetic variants in clones, detected as differences from  wild-type  cells, is presented as a "gt" table. This table classifies the biotypes characterizing protein-coding genes, pseudogenes, and noncoding elements.

```{r}
#Total count###########
ClonesTable<-Clones1 %>% 
  mutate(Biotype = case_when(
    Biotype == "noncoding, pseudogene" ~ "pseudogene, noncoding",
    Biotype == "pseudogene, protein_coding" ~ "protein_coding, pseudogene",
    Biotype == "pseudogene, protein_coding, noncoding" ~ "protein_coding, pseudogene, noncoding",
    TRUE ~ as.character(Biotype)
  )) %>%
  count(Biotype, CLONE) %>%
  arrange(factor(CLONE, levels = c("II21", "IX44", "VII11", "A13", "A40"))) 
```

```{r}
Biotype <- ClonesTable %>%
  pivot_wider(
    names_from = CLONE,
    values_from = n,
    values_fn = mean
  )
```
 Creating "gt" table
```{r}
Biotype_gt <- gt(Biotype, rowname_col = "Biotype") %>%
  grand_summary_rows(
    fns = list(label = md("**TOTAL**"), id = "totals", fn = "sum"),
    side = "bottom"
  ) %>%
  tab_style(
    style = cell_fill(color = "lightblue" |> adjust_luminance(steps = +1)),
    locations = cells_grand_summary()
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_grand_summary(columns = c("Biotype", "II21", "IX44", "VII11"))
  ) %>%
  tab_style(
    style = cell_text(align = "right"),
    locations = cells_grand_summary(columns = c("A13", "A40"))
  ) %>%
  tab_style(
    style = cell_fill(color = "azure1"),
    locations = cells_body(columns = c("Biotype", "II21", "IX44", "VII11", "A13", "A40"))
  ) %>%
  tab_spanner(
    label = md("*NOVOGENE CORPORATION INC*"),
    columns = c("II21", "IX44", "VII11"),
    id = "dt1"
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_column_spanners(spanners = "dt1")
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_column_spanners(spanners = "dt1")
  ) %>%
  tab_spanner(
    label = md("*NORTHWESTERN*"),
    columns = c("A13", "A40"),
    id = "dt2"
  ) %>%
  tab_style(
    style = cell_text(align = "right"),
    locations = cells_column_spanners(spanners = "dt2")
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_column_spanners(spanners = "dt2")
  ) %>%
  tab_header(
    title = md("**Number of spontaneous variants in clones detected as different from ME180 wt  cells**"),
    subtitle = md("*All chromosomes*")
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_title(groups = "subtitle")
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_body(columns = c("Biotype", "II21", "IX44", "VII11"))
  ) %>%
  tab_style(
    style = cell_text(align = "right"),
    locations = cells_column_labels(columns = c("A13", "A40"))
  ) %>%
  tab_style(
    style = cell_text(align = "right"),
    locations = cells_body(columns = c("A13", "A40"))
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_column_labels(columns = c("II21", "IX44", "VII11"))
  ) %>%
  opt_footnote_marks(marks = "standard") %>%
  tab_footnote(
    footnote = md("*VII11 is the parent clone for A13 and A40.*"),
    locations = cells_column_labels(columns = VII11),
    placement = "right"
  )
#print table
Biotype_gt

```

Overall, comparisons between genomic sequences of clones and wild type cells revealed a significant number of genomic variants belonging to different sequence categories (noncoding, protein coding and pseudogenes inclusive of their mutual overlap). Majority of these sequence variants were not a direct consequence of CRIPR/Cas9 editing. 

### Venn diagram of sequence variants in all three sequence categories: noncoding, protein coding and pseudogenes, including overlapping positions

To analyze and visualize the overlap of genomic variants across different clones we used Venn diagrams. R code to create these diagrams utilized the `ggVennDiagram` package and other supporting libraries. First, the "Clones" dataset was  filtered to include only entries where "Biotype" matches "protein_coding".
```{r}
ClonesVenn <- Clones%>%
  filter(Biotype == "protein_coding")%>%
  select("CHROM_POS_Clone","CLONE","Biotype","CHROM_POS")%>%
  group_by(CHROM_POS_Clone) %>%
  summarise_all(~toString(unique(.))) 

```

```{r}
ClonesVenn13<-ClonesVenn%>%
  filter(CLONE=="VII11" |CLONE=="IX44"|CLONE=="II21" )

ClonesVenn23<-ClonesVenn%>%
  filter(CLONE=="VII11" |CLONE=="A13"|CLONE=="A40" )


x13 <- split(ClonesVenn13$CHROM_POS, ClonesVenn13$CLONE)
gg_venn13 <- ggVennDiagram(x13, label_alpha = 0, label_size = 3, set_color = "black", edge_size = 0.4) +
  scale_fill_gradient(low = "white", high = "darkslategray2", name = "Color") +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("Protein coding")

x23 <- split(ClonesVenn23$CHROM_POS, ClonesVenn23$CLONE)
gg_venn23 <- ggVennDiagram(x23, label_alpha = 0, label_size = 3, set_color = "black", edge_size = 0.4) +
  scale_fill_gradient(low = "white", high = "darkseagreen1", name = "Color") +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("Protein coding")


# Arrange the second row of plots
# Create the combined text grob
top_grob <- textGrob("A                                                                B", x = 0.02, hjust = 0.02, gp = gpar(fontsize = 14, fontface = "bold"))

# Arrange the Venn diagrams
Venn <- grid.arrange(gg_venn13, gg_venn23, ncol = 2, top = top_grob)
```

By standardizing "Biotype" labels, filtering relevant data, and creating comprehensive plots, we illustrated the differences and similarities between genomes of clones compared to genome of ME180 wild-type cells. 

### The analyzis of the sequence variants classified as High impact effect 

Counting of High impact effect variants in clones VII11, A13 and A40 was done by filtering  "Effect_Impact" and "CLONE" columns.
```{r}
High_Impact<-Clones%>% filter(Effect_Impact=="HIGH")
High_Impact <- High_Impact %>% filter(CLONE %in% c("VII11", "A13", "A40"))
chor <- ftable(High_Impact$CLONE, High_Impact$GENE) %>%
  as_tibble()
```

The data visualization was performed by `chordDiagram()` function.
```{r}
#data preparation
chor <- ftable(High_Impact$CLONE, High_Impact$GENE) %>%
  as_tibble()

```
```{r}

chordDiagram(chor, 
             annotationTrack = "grid", 
             preAllocateTracks = list(track.height = max(strwidth(unlist(dimnames(chor))))),
             col = c("magenta", "springgreen", "royalblue2"))  # Specify desired colors

# Customize grid appearance by adding a custom track
circos.track(track.index = 1, panel.fun = function(x, y) {
  circos.axis(labels.cex = 0.6, labels.col = "gray", major.tick.length = 0.2, col = "gray")
}, bg.border = NA)

# Customize sector labels
circos.track(track.index = 1, panel.fun = function(x, y) {
  circos.text(CELL_META$xcenter, CELL_META$ylim[1], CELL_META$sector.index, 
              facing = "clockwise", cex = 0.7, niceFacing = FALSE, adj = c(0, 0.5))
}, bg.border = NA)

# Clear the circos plot
circos.clear()
```

Sequence variants classified as having High impact effect were found in 52 genes (in addition to EGFR) in clones A13, A40 and VII11. 

```{r}
STEP1<-High_Impact%>%
  select("GENE" ,"CHROM", "POS","REF" ,"ALT","Effect","Rank","HGVS_cDNA", "HGVS_Protein","CLONE")
STEP2 <- STEP1%>%
  as_tibble()%>%
  mutate("All_inf"=paste(GENE, CHROM, POS, REF, ALT, Effect,Rank,HGVS_cDNA, HGVS_Protein,sep="="))
STEP3 <- ftable(STEP2$All_inf, STEP2$CLONE) %>%
  as_tibble() %>%
  pivot_wider(
    names_from = Var2,
    values_from = Freq,
    values_fn = mean
  ) %>%
  mutate(across(c("A13", "A40", "VII11"), ~recode(., `1` = "✔️", `0` = "✘"))) #✔, 
STEP4<- separate(STEP3,col=Var1,into=c("GENE" ,"CHROM", "POS","REF" ,"ALT","Effect","Rank","HGVS_cDNA", "HGVS_Protein"),sep="=",remove=T)
View(STEP4)
```

```{r}
Gene_number<-ftable(STEP4$GENE)
View(Gene_number)
```


```{r}
Hight_gt <- STEP4 %>%
  mutate("REF➩ALT" = paste(REF, ALT, sep = "➩")) %>%
  select("GENE" ,"CHROM", "POS", "REF➩ALT", "Effect","Rank","HGVS_cDNA", "HGVS_Protein","VII11", "A13", "A40",) %>%
  gt(rowname_col =  "GENE",
     groupname_col = "CHROM",
     row_group_as_column = F,
     rownames_to_stub = F,
     process_md = TRUE,
     row_group.sep = getOption("gt.row_group.sep", " - ")
  ) |>
  tab_spanner(
    label = "CRISPR CLONES",
    columns = c("VII11","A13", "A40"),
    id = "dt"
  ) |>
  tab_style(
    style = cell_fill(color = "#C1CDCD"),
    locations = cells_body(columns = c("VII11","A13", "A40"))
    )|>
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_body(
      columns = c("GENE" ,"CHROM", "POS", "REF➩ALT", "Effect","Rank","HGVS_cDNA", "HGVS_Protein"))
   )|>
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_column_spanners(spanners = "dt")
  )|>
  tab_header(
    title = md("**High effect impact mutations in clones compared to ME-180 wild type sequence**")
  )|>
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_column_labels(columns = everything())
  )
  

Hight_gt 
```


### Analysis of potential off-target mutations generated by action of Cas9 RNA-guided endonuclease 

The analysis of potential off-target sites for Cas9 RNA-guided endonucleases was conducted for the two guide RNAs (Y45-2 and L14-1), allowing up to four mismatches or gaps, using the  algorithm Cas-OFFinder [CRISPR RGEN Tools](http://www.rgenome.net/cas-offinder/). 

```{r}
L14_1<-read.csv("L14 1.txt")|>
  as.data.frame()|>
  mutate("SENSE"="L14_1")

Y45_2<-read.csv("Y45 2.txt")|>
  as.data.frame()|>
  mutate("SENSE"="Y45_2")
```

It is important to note that the gRNA had no off-target sequences with less than 3 mismatches. Full matches were found only for the two intended  positions: chr7, position 55142299 for L14-1 (AGTAACAAGCTCACGCAGTTGGG) and position 55142390 for Y45-2 (TGGAAATTACCTATGTGCAGAGG).

```{r}
L14_1print <- L14_1 %>%
  as_tibble() %>%
  slice(14) %>%
  print()

```

```{r}
Y45_2print <- Y45_2 %>%
  as_tibble() %>%
  slice(75) %>%
  print()
```

Both gRNA sequences fully matched the EGFR protein genomic sequence. In search for sequence matches an extra nucleotide was added at the beginning (for direct complement) or at the end (for reverse complement) of the sequence. For example, position 55142299 includes an extra nucleotide “G” at the beginning (“**G**”AGTAACAAGCTCACGCAGTTGGG).
Based on the Cas-OFFinder output files, there were 23 mismatched targets in the genome; positions of these mismatched targets, were compared with the DNA positions of sequence variants detected in clones as different from ME180 wild type sequence.
```{r}
off_target<-as_tibble(rbind(L14_1,Y45_2))
off_target <- separate(off_target, 
                       col = "X.Bulge.type.crRNA.DNA.Chromosome.Position.Direction.Mismatches.Bulge.Size",
                       into = c("Bulge_type", "crRNA", "DNA", "Chromosome", "Position", "Direction", "Mismatches", "Bulge_Size"),
                       sep = "\t",
                       remove = T)
OFF_target <- off_target %>%mutate(COMPLEMEN = case_when(Direction == "+" ~ "direct", TRUE ~ "reverse "))%>%
  mutate(Position = as.double(Position))%>%
  mutate(START = case_when(Direction == "+" ~  Position, TRUE ~ Position+24))%>%
  mutate(END = case_when(Direction == "+" ~  Position+24, TRUE ~ Position))%>%
  relocate(SENSE,.before = Chromosome)%>%
  select("SENSE","CHROM"="Chromosome","Position","Direction","Mismatches","COMPLEMEN","START", "END")
#Comparison data#################

Comparison<-filter(Clones1,CHROM%in%OFF_target$CHROM)
Comparison1<-merge(Comparison,OFF_target, by = "CHROM", all.x = TRUE)
Comparison2 <- Comparison1 %>%
  filter((Direction == "+" & POS >= START & POS <= END) |
           (Direction == "-" & POS <= START & POS >= END))%>%
  select("SENSE","CHROM", "GENE","Effect","POS","REF","ALT","Direction", "COMPLEMEN","Mismatches","CLONE")
Comparison3 <- Comparison2%>%
  as_tibble()%>%
  mutate("All_inf"=paste(SENSE,CHROM,GENE,Effect,POS,REF,ALT,Direction,COMPLEMEN,Mismatches,sep="="))
Comparison4 <- ftable(Comparison3$All_inf, Comparison3$CLONE) %>%
  as_tibble() %>%
  pivot_wider(
    names_from = Var2,
    values_from = Freq,
    values_fn = mean
  ) %>%
  mutate(across(c("A13", "A40", "II21", "IX44", "VII11"), ~recode(., `1` = "✔️", `0` = "✘"))) #✔, 
Comparison5<- separate(Comparison4,col=Var1,into=c("SENSE","CHROM","GENE","Effect","POS","REF","ALT","Direction","COMPLEMEN","Mismatch_n"),sep="=",remove=T)%>%
  mutate("REF_ALT" = paste(REF, ALT, sep = "➩"))

```

Creating "gt" table
```{r}
Comparison5_gt <- Comparison5 %>%
  mutate("REF_MUT" = paste(REF, ALT, sep = "➩")) %>%
  select("SENSE", "CHROM", "POS", "GENE", "EFFECT"=Effect, "REF_MUT", "A13", "A40", "VII11", "II21", "IX44") %>%
  gt(rowname_col = "CHROM",
     groupname_col = "SENSE",
     row_group_as_column = F,
     rownames_to_stub = F,
     process_md = TRUE,
     row_group.sep = getOption("gt.row_group.sep", " - ")
  ) %>%
  tab_spanner(
    label = "CRISPR CLONES",
    columns = c("A13", "A40", "VII11", "II21", "IX44"),
    id = "dt"
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_body(columns = c("CHROM", "POS", "GENE", "EFFECT", "REF_MUT"))
  ) %>%
  tab_style(
    style = cell_text(weight = "bold"),
    locations = cells_column_spanners(spanners = "dt")
  ) %>%
  tab_header(
    title = md("**Analysis of DNA variant positions detected as different from ME-180wt based on potential off-target effects of Cas9 RNA-guided endonucleases**"),
    subtitle = md("<div style='text-align: left'>L14-1(sense):crRNA: CRISPR813570 (AGTAACAAGCTCACGCAGTT, PAM seq GGG)<br>Y45-2 (sense): crRNA: TGGAAATTACCTATGTGCAG, PAM seq is AGG</div>")
  ) %>%
  tab_source_note(
    source_note = md("**Source: CRISPR RGEN Tools:**  *http://www.rgenome.net/cas-offinder/*")
  ) %>%
  tab_source_note(
    source_note = md("*at the maximum Mismatch Number= 4* ")
  ) %>%
  tab_style(
    style = cell_text(align = "left"),
    locations = cells_column_labels(columns = everything())
  )

Comparison5_gt
```

No mutations in clones were found to be an outcome of the off-target editing events caused by guide RNA. The only mutations associated with use of gRNAs were those related to the EGFR gene.

# Session Info:

```{r}
sessionInfo()
```
